__tests__/Auth0Client.test.ts

```diff
@@ -173,28 +173,389 @@ describe('Auth0Client', () => {
+        code_challenge_method: 'S256'
+      });
+    });
+
+    it('should log the user in with a popup and redirect using a default redirect URI', async () => {
+      const auth0 = setup({ leeway: 10, redirect_uri: null });
+
+      await loginWithPopup(auth0, {
+        connection: 'test-connection',
+        audience: 'test'
+      });
+
+      expect(mockWindow.open).toHaveBeenCalled();
+      const [[url]] = (<jest.Mock>mockWindow.open).mock.calls;
+      assertUrlEquals(url, 'auth0_domain', '/authorize', {
+        redirect_uri: 'http://localhost',
+        client_id: TEST_CLIENT_ID,
+        scope: TEST_SCOPES,
+        response_type: 'code',
+        response_mode: 'web_message',
+        state: TEST_STATE,
+        nonce: TEST_NONCE,
+        code_challenge: TEST_CODE_CHALLENGE,
+        code_challenge_method: 'S256',
+        connection: 'test-connection',
+        audience: 'test'
+      });
+    });
+
+    it('should log the user in with a popup and redirect', async () => {
+      const auth0 = setup({ leeway: 10 });
+
+      await loginWithPopup(auth0, {
+        connection: 'test-connection',
+        audience: 'test'
+      });
+
+      expect(mockWindow.open).toHaveBeenCalled();
+      const [[url]] = (<jest.Mock>mockWindow.open).mock.calls;
+      assertUrlEquals(url, TEST_DOMAIN, '/authorize', {
+        redirect_uri: TEST_REDIRECT_URI,
+        client_id: TEST_CLIENT_ID,
+        scope: TEST_SCOPES,
+        response_type: 'code',
+        response_mode: 'web_message',
+        state: TEST_STATE,
+        nonce: TEST_NONCE,
+        code_challenge: TEST_CODE_CHALLENGE,
+        code_challenge_method: 'S256',
+        connection: 'test-connection',
+        audience: 'test'
+      });
+    });
+
+    it('should log the user in with a popup and get the token', async () => {
+      const auth0 = setup();
+
+      await loginWithPopup(auth0);
+
+      expect(mockWindow.open).toHaveBeenCalled();
+      assertPost('https://auth0_domain/oauth/token', {
+        redirect_uri: TEST_REDIRECT_URI,
+        client_id: TEST_CLIENT_ID,
+        code_verifier: TEST_CODE_VERIFIER,
+        grant_type: 'authorization_code',
+        code: 'my_code'
+      });
+    });
+
+    it('uses default config', async () => {
+      const auth0 = setup({ leeway: 10 });
+
+      await loginWithPopup(auth0);
+
+      expect(utils.runPopup).toHaveBeenCalledWith(
+        expect.any(String),
+        DEFAULT_POPUP_CONFIG_OPTIONS
+      );
+    });
+
+    it('should be able to provide custom config', async () => {
+      const auth0 = setup({ leeway: 10 });
+
+      await loginWithPopup(auth0, {}, { timeoutInSeconds: 3 });
+
+      expect(utils.runPopup).toHaveBeenCalledWith(expect.any(String), {
+        timeoutInSeconds: 3
+      });
+    });
+
+    it('throws an error if not resolved before timeout', async () => {
+      const auth0 = setup({ leeway: 10 });
+
+      await expect(
+        loginWithPopup(auth0, {}, { timeoutInSeconds: 0.005 })
```

Comment:
To keep the tests fast, the popup closes in 10 miliseconds. This means we need to set the timeout lower to make it timeout.